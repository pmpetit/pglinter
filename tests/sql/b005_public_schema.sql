-- Regression test for B005 rule: Schemas with public CREATE privileges
-- Tests the percentage-based detection of schemas allowing CREATE for public role
-- versus total schemas in the database

BEGIN;

-- Create the extension
DROP EXTENSION IF EXISTS pglinter CASCADE;
CREATE EXTENSION IF NOT EXISTS pglinter;

SELECT 'B005 Regression Test: Schemas with public CREATE privileges' AS test_header;

-- Setup B005 rule for testing
SELECT pglinter.disable_all_rules();
SELECT pglinter.enable_rule('B005');

-- PART 1: Test with LOW percentage of insecure schemas (should NOT trigger)
SELECT 'PART 1: Testing with LOW percentage of insecure schemas (should NOT trigger)' AS test_part;

-- Create one secure schema (no public CREATE)
CREATE SCHEMA test_secure_schema;
REVOKE CREATE ON SCHEMA test_secure_schema FROM public;

-- Create one secure schema
CREATE SCHEMA test_secure_schema_1;
REVOKE CREATE ON SCHEMA test_secure_schema_1 FROM public;

-- Create one secure schema
CREATE SCHEMA test_secure_schema_2;
REVOKE CREATE ON SCHEMA test_secure_schema_2 FROM public;

-- Create one secure schema
CREATE SCHEMA test_secure_schema_3;
REVOKE CREATE ON SCHEMA test_secure_schema_3 FROM public;

-- Create one secure schema
CREATE SCHEMA test_secure_schema_4;
REVOKE CREATE ON SCHEMA test_secure_schema_4 FROM public;

-- Create one secure schema
CREATE SCHEMA test_secure_schema_5;
REVOKE CREATE ON SCHEMA test_secure_schema_5 FROM public;

-- Create one secure schema
CREATE SCHEMA test_secure_schema_6;
REVOKE CREATE ON SCHEMA test_secure_schema_6 FROM public;

-- Create one secure schema
CREATE SCHEMA test_secure_schema_7;
REVOKE CREATE ON SCHEMA test_secure_schema_7 FROM public;

-- Create one secure schema
CREATE SCHEMA test_secure_schema_8;
REVOKE CREATE ON SCHEMA test_secure_schema_8 FROM public;


-- Create one insecure schema
CREATE SCHEMA test_insecure_schema_1;
GRANT CREATE ON SCHEMA test_insecure_schema_1 TO public;


-- Test B005 with low percentage (should not trigger with default thresholds)
SELECT 'Running B005 check with LOW percentage of insecure schemas (should not trigger with default thresholds):' AS test_1;
SELECT pglinter.perform_base_check();

CREATE SCHEMA test_insecure_schema_2;
GRANT CREATE ON SCHEMA test_insecure_schema_2 TO public;

-- Test B005 with high percentage (should trigger)
SELECT 'Running B005 check with HIGH percentage of insecure schemas (should trigger):' AS test_2;
SELECT pglinter.perform_base_check();
-- Test with file output
SELECT pglinter.perform_base_check('/tmp/pglinter_b005_results.sarif');
-- Test if file exists and show checksum
\! md5sum /tmp/pglinter_b005_results.sarif


-- PART 3: Test threshold adjustment
SELECT 'PART 3: Testing B005 threshold adjustments (should not trigger)' AS test_part;

-- Lower the warning threshold to ensure detection
SELECT pglinter.update_rule_levels('B005', 50, 80);

SELECT 'B005 thresholds updated to warning=50%, error=80%' AS threshold_info;

-- Test with adjusted thresholds
SELECT 'Running B005 check with adjusted thresholds (warning=50%):' AS test_3;
SELECT pglinter.perform_base_check();

-- PART 6: Verification of SQL queries used by B005
SELECT 'PART 6: Direct verification of B005 SQL queries' AS sql_verification;

-- Test rule explanation
SELECT 'B005 rule explanation:' AS rule_explanation;
SELECT pglinter.explain_rule('B005');

ROLLBACK;
