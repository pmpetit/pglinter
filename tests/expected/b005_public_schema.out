-- Regression test for B005 rule: Schemas with public CREATE privileges
-- Tests the percentage-based detection of schemas allowing CREATE for public role
-- versus total schemas in the database
BEGIN;
-- Create the extension
DROP EXTENSION IF EXISTS pglinter CASCADE;
CREATE EXTENSION IF NOT EXISTS pglinter;
SELECT 'B005 Regression Test: Schemas with public CREATE privileges' as test_header;
                         test_header                         
-------------------------------------------------------------
 B005 Regression Test: Schemas with public CREATE privileges
(1 row)

-- Setup B005 rule for testing
SELECT pglinter.disable_all_rules();
NOTICE:  üî¥ Disabled 22 rule(s)
 disable_all_rules 
-------------------
                22
(1 row)

SELECT pglinter.enable_rule('B005');
NOTICE:  ‚úÖ Rule B005 has been enabled
 enable_rule 
-------------
 t
(1 row)

-- PART 1: Test with LOW percentage of insecure schemas (should NOT trigger)
SELECT 'PART 1: Testing with LOW percentage of insecure schemas (should NOT trigger)' as test_part;
                                  test_part                                   
------------------------------------------------------------------------------
 PART 1: Testing with LOW percentage of insecure schemas (should NOT trigger)
(1 row)

-- Create one secure schema (no public CREATE)
CREATE SCHEMA test_secure_schema;
REVOKE CREATE ON SCHEMA test_secure_schema FROM PUBLIC;
-- Create one secure schema
CREATE SCHEMA test_secure_schema_1;
REVOKE CREATE ON SCHEMA test_secure_schema_1 FROM PUBLIC;
-- Create one secure schema
CREATE SCHEMA test_secure_schema_2;
REVOKE CREATE ON SCHEMA test_secure_schema_2 FROM PUBLIC;
-- Create one secure schema
CREATE SCHEMA test_secure_schema_3;
REVOKE CREATE ON SCHEMA test_secure_schema_3 FROM PUBLIC;
-- Create one secure schema
CREATE SCHEMA test_secure_schema_4;
REVOKE CREATE ON SCHEMA test_secure_schema_4 FROM PUBLIC;
-- Create one secure schema
CREATE SCHEMA test_secure_schema_5;
REVOKE CREATE ON SCHEMA test_secure_schema_5 FROM PUBLIC;
-- Create one secure schema
CREATE SCHEMA test_secure_schema_6;
REVOKE CREATE ON SCHEMA test_secure_schema_6 FROM PUBLIC;
-- Create one secure schema
CREATE SCHEMA test_secure_schema_7;
REVOKE CREATE ON SCHEMA test_secure_schema_7 FROM PUBLIC;
-- Create one secure schema
CREATE SCHEMA test_secure_schema_8;
REVOKE CREATE ON SCHEMA test_secure_schema_8 FROM PUBLIC;
-- Create one insecure schema
CREATE SCHEMA test_insecure_schema_1;
GRANT CREATE ON SCHEMA test_insecure_schema_1 TO PUBLIC;
-- Test B005 with low percentage (should not trigger with default thresholds)
SELECT 'Running B005 check with LOW percentage of insecure schemas (should not trigger with default thresholds):' as test_1;
                                                  test_1                                                  
----------------------------------------------------------------------------------------------------------
 Running B005 check with LOW percentage of insecure schemas (should not trigger with default thresholds):
(1 row)

SELECT pglinter.perform_base_check();
NOTICE:  ‚úÖ No issues found - database schema looks good!
 perform_base_check 
--------------------
 t
(1 row)

CREATE SCHEMA test_insecure_schema_2;
GRANT CREATE ON SCHEMA test_insecure_schema_2 TO PUBLIC;
-- Test B005 with high percentage (should trigger)
SELECT 'Running B005 check with HIGH percentage of insecure schemas (should trigger):' as test_2;
                                    test_2                                     
-------------------------------------------------------------------------------
 Running B005 check with HIGH percentage of insecure schemas (should trigger):
(1 row)

SELECT pglinter.perform_base_check();
NOTICE:  üîç pglinter found 1 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ö†Ô∏è  [B005] WARNING: 3/12 schemas are unsecured, schemas where all users can create objects in, exceed the warning threshold: 25%.
NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 1 warning(s), 0 info
NOTICE:  üü° Some warnings found - consider reviewing for optimization
 perform_base_check 
--------------------
 t
(1 row)

-- PART 3: Test threshold adjustment
SELECT 'PART 3: Testing B005 threshold adjustments (should not trigger)' as test_part;
                            test_part                            
-----------------------------------------------------------------
 PART 3: Testing B005 threshold adjustments (should not trigger)
(1 row)

-- Lower the warning threshold to ensure detection
SELECT pglinter.update_rule_levels('B005', 50, 80);
NOTICE:  ‚úÖ Updated rule B005 levels: warning=50, error=80
 update_rule_levels 
--------------------
 t
(1 row)

SELECT 'B005 thresholds updated to warning=50%, error=80%' as threshold_info;
                  threshold_info                   
---------------------------------------------------
 B005 thresholds updated to warning=50%, error=80%
(1 row)

-- Test with adjusted thresholds
SELECT 'Running B005 check with adjusted thresholds (warning=50%):' as test_3;
                           test_3                           
------------------------------------------------------------
 Running B005 check with adjusted thresholds (warning=50%):
(1 row)

SELECT pglinter.perform_base_check();
NOTICE:  ‚úÖ No issues found - database schema looks good!
 perform_base_check 
--------------------
 t
(1 row)

-- PART 6: Verification of SQL queries used by B005
SELECT 'PART 6: Direct verification of B005 SQL queries' as sql_verification;
                sql_verification                 
-------------------------------------------------
 PART 6: Direct verification of B005 SQL queries
(1 row)

-- Test rule explanation
SELECT 'B005 rule explanation:' as rule_explanation;
    rule_explanation    
------------------------
 B005 rule explanation:
(1 row)

SELECT pglinter.explain_rule('B005');
NOTICE:  üìñ Rule Explanation for B005
============================================================

üéØ Rule Name: UnsecuredPublicSchema
üìã Scope: BASE

üìù Description:
Only authorized users should be allowed to create objects.

‚ö†Ô∏è  Message Template:
{0}/{1} schemas are unsecured, schemas where all users can create objects in, exceed the {2} threshold: {3}%.

üîß How to Fix:
   1. REVOKE CREATE ON SCHEMA <schema_name> FROM PUBLIC
============================================================
 explain_rule 
--------------
 t
(1 row)

ROLLBACK;
